homeassistant/components/simplisafe/__init__.py

```diff
@@ -474,6 +474,15 @@ async def async_init(self):
                 _LOGGER.error("Error while fetching initial event: %s", err)
                 self.initial_event_to_use[system.system_id] = {}
 
+            # The initial update doesn't process notifications because it is likely that
+            # they will trigger SIMPLISAFE_NOTIFICATION events before other critical
+            # components – such as automation – are ready. Therefore, we schedule an
+            # update of notifications as soon as HASS is fully started:
+            self._hass.bus.async_listen_once(
+                EVENT_HOMEASSISTANT_START,
+                lambda event: self._async_process_new_notifications(system),
```

Comment:
I think I've found it. The lambda is interpreted as to be run in sync context. Use a regular function and decorate it with `@callback` instead.

Reply:
@MartinHjelmare I don't think that's the whole story, unfortunately. If I change things to look like this:

```python
@callback
def process_notifications(_):
    self._async_process_new_notifications(system)

self._hass.bus.async_listen_once(
    EVENT_HOMEASSISTANT_START, process_notifications
)
```

...I still frequently see events being fired before the `automation` integration is fully set up:

```txt
2020-04-15 08:20:22 DEBUG (MainThread) [homeassistant.components.simplisafe] New system notifications: {SystemNotification(notification_id='c17097c3-04b2-4a99-b064-244ae334431b', text='Alarm Cancelled by Master', category='alarm', code='1001', timestamp=datetime.datetime(2020, 4, 14, 3, 13, 18, tzinfo=<UTC>), link='', link_label=''), SystemNotification(notification_id='978cc8e4-bd9e-41e0-a583-ed2acbdef6dd', text='Keypad Panic "Kitchen" triggered', category='alarm', code='1000', timestamp=datetime.datetime(2020, 4, 14, 3, 13, 7, tzinfo=<UTC>), link='', link_label='')}
2020-04-15 08:20:22 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event SIMPLISAFE_NOTIFICATION[L]: category=alarm, code=1001, message=Alarm Cancelled by Master, timestamp=2020-04-13T21:13:18-06:00>
2020-04-15 08:20:22 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event SIMPLISAFE_NOTIFICATION[L]: category=alarm, code=1000, message=Keypad Panic "Kitchen" triggered, timestamp=2020-04-13T21:13:07-06:00>
2020-04-15 08:20:22 INFO (MainThread) [homeassistant.components.automation] Initialized trigger Notify on SimpliSafe events
```

This yields the same challenge of automations not processing the events.

Reply:
At least the events are now fired in the event loop like they should.

Will think some more about this.

Reply:
Can you include the home assistant start event logging line just for completeness?

Reply:
Yes sir, updated my logs above. I removed other lines that weren't relevant, FYI (SSDP, etc.).

Reply:
a0423d2 seems to do the trick; thoughts?

@balloob: @MartinHjelmare and I were chatting on this; might it make sense for a new event that is fired after automation triggers are fully set up (post-`homeassistant_start`) to help this kind of use case?

Reply:
We just added a similar startup stage to shutdown (final_write), we can do this at startup too. However in this case, aren't these events old? Should they even be fired to begin with? How long will simplisafe keep events around if we are offline, could we end up firing a 2 hour old event ?

Reply:
SimpliSafe will keep notifications in place until they are manually cleared in the SimpliSafe web app (currently thinking about whether HASS should influence that longer term).

Yes, they should be fired regardless of age. Example: if I get a notification that I have a SimpliSafe firmware upgrade while HASS is down/offline, I would still want to be notified of it when HASS restarts.